test/e2e/firewall.go

```diff
@@ -0,0 +1,410 @@
+	count := 0
+
+	framework.Logf("Waiting up to %v for not hitting %v for %v times", timeout, targetHost, countToSucceed)
+	condition := func() (bool, error) {
+		respBody, err := queryFromUrl(externalIP, "hostname", httpPort, firewallTestHttpTimeout)
+		if err != nil {
+			framework.Logf("%v", err)
+			return false, nil
+		}
+		hittedHost := strings.TrimSpace(respBody)
+		if hittedHost == targetHost {
+			framework.Logf("Error hitting unexpected host: %v, reset counter: %v", targetHost, count)
+			count = 0
+		} else {
+			count++
+		}
+
+		if count >= countToSucceed {
+			return true, nil
+		}
+		return false, nil
+	}
+
+	if err := wait.Poll(time.Second, timeout, condition); err != nil {
+		return fmt.Errorf("error: not hitting %v for %v times, expect at least %v times, err message: %v", targetHost, count, countToSucceed, err)
+	}
+	return nil
+}
+
+func queryFromUrl(ip, path string, port int32, timeout time.Duration) (string, error) {
+	// Create a new Transport for not re-using connnections
+	client := &http.Client{
+		Timeout:   timeout,
+		Transport: &http.Transport{},
+	}
+	url := fmt.Sprintf("http://%v:%v/%v", ip, port, path)
+	resp, err := client.Get(url)
+	if err != nil {
+		return "", err
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != 200 {
+		return "", fmt.Errorf("Got non-ok status code: %v", resp.StatusCode)
+	}
+
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return "", err
+	}
+	return string(body), nil
+}
+
+func setNodeTags(gceCloud *gcecloud.GCECloud, projectID, zone string, nodeName string, tags []string) ([]string, error) {
+	// Re-get instance everytime because we need the latest fingerprint for updating metadata
+	res, err := gceCloud.GetComputeService().Instances.Get(projectID, zone, nodeName).Do()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get instance: %v", err)
+	}
+
+	_, err = gceCloud.GetComputeService().Instances.SetTags(projectID, zone, nodeName, &compute.Tags{Fingerprint: res.Tags.Fingerprint, Items: tags}).Do()
+	if err != nil {
+		return nil, fmt.Errorf("failed to set instance tags: %v", err)
+	}
+	framework.Logf("Sent request to set tags %v on instance: %v", tags, nodeName)
+	if err := waitForNodeTags(gceCloud, projectID, zone, nodeName, tags, firewallTimeoutDefault); err != nil {
+		return res.Tags.Items, err
+	}
+	framework.Logf("Saw expected tags on instance.")
+	return res.Tags.Items, nil
+}
+
+func waitForNodeTags(gceCloud *gcecloud.GCECloud, projectID, zone string, nodeName string, tags []string, timeout time.Duration) error {
+	var tagsFromGCE []string
+	condition := func() (bool, error) {
+		res, err := gceCloud.GetComputeService().Instances.Get(projectID, zone, nodeName).Do()
+		if err != nil {
+			framework.Logf("Failed to get instance: %v", err)
+			return false, nil
+		}
+		tagsFromGCE = res.Tags.Items
+		if err := sameStringElements(tagsFromGCE, tags); err != nil {
```

Comment:
sameStringElements is a slightly confusing name, suggest a rename + docstring to explain what the purpose of the function is